src/amp/lib/experiment.ts

```diff
@@ -0,0 +1,150 @@
+        acc = { ...acc, ...{ [variantName]: config[fieldName] } };
+        return acc;
+    }, {});
+};
+
+const isValidAmpModelVariant = (variant: {
+    [variantName: string]: string | number;
+}): variant is { [variantName: string]: number } => {
+    return Object.values(variant).every(x => typeof x === 'number');
+};
+
+const isValidStyleVariant = (variant: {
+    [variantName: string]: string | number;
+}): variant is { [variantName: string]: string } => {
+    return Object.values(variant).every(x => typeof x === 'string');
+};
+
+export const extractExperimentModels = (
+    fullConfig: StyledModelCollection,
+): AmpModelCollection => {
+    return Object.entries(fullConfig).reduce(
+        (extractedModels, [experimentName, styledModel]) => {
+            const { variants, ...configWithoutVariants } = styledModel;
+            const proportions = extractFromVariants(variants, 'proportion');
+            if (isValidAmpModelVariant(proportions)) {
+                extractedModels[experimentName] = {
+                    ...configWithoutVariants,
+                    ...{
+                        variants: proportions,
+                    },
+                };
+            }
+            return extractedModels;
+        },
+        {} as AmpModelCollection,
+    );
+};
+
+export const extractExperimentStyles = (
+    fullConfig: StyledModelCollection,
+): StylesCollection => {
+    return Object.entries(fullConfig).reduce(
+        (extractedModels, [experimentName, styledModel]) => {
+            const { variants } = styledModel;
+            const styles = extractFromVariants(variants, 'style');
+            if (isValidStyleVariant(styles)) {
+                extractedModels[experimentName] = { ...styles };
+            }
+            return extractedModels;
+        },
+        {} as StylesCollection,
+    );
+};
+
+export const getActiveExperiments = (
+    abTestObject: StyledModelCollection,
+    switches: { [key: string]: boolean },
+): StyledModelCollection => {
+    return Object.entries(abTestObject).reduce(
+        (acc, [testName, testConfig]) => {
+            if (switches[testName]) {
+                acc[testName] = testConfig;
+            }
+            return acc;
+        },
+        {} as StyledModelCollection,
+    );
+};
+
+export const buildExperimentStyle = (experiments: StylesCollection): string => {
+    return Object.entries(experiments).reduce(
+        (experimentStyleString, [experimentName, styleVariants]) => {
+            const experimentStyle = Object.entries(styleVariants).reduce(
+                (variantStyleString, [variantName, variantStyle]) => {
+                    return `${variantStyleString} body[amp-x-${experimentName}='${variantName}'] ${variantStyle}`.trim();
+                },
+                '',
+            );
+            return `${experimentStyleString} ${experimentStyle}`.trim();
+        },
+        '',
+    );
+};
+
+// export const extractProportionFromVariants = (variants: {
```

Comment:
why is this commented out?

Reply:
There were two approaches to this - one using a single `extractFromVariants` with type predicates and one with two almost identical `extract...FromVariants` functions. I wanted to check which would be preferable at this stage and then remove the one that isn't in use.

Which approach do you think is better here?

Reply:
I think the commented out section seems more readable, it uses reduce more sparingly and is easier to follow, so I would favor that if possible